backend/alternative_flights/alt_flight_search.py

```diff
@@ -0,0 +1,232 @@
+            if current_option.total_duration <= 24:  # Ensure total trip time is <= 24 hours
+                options.append(current_option)
+            return
+
+        for flight in filtered_data.get(current_airport, []):
+            if len(current_option.flights) == 0 and flight.departure_time < current_time:
+                continue
+            elif len(current_option.flights) > 0 and flight.departure_time < current_time + timedelta(hours=1):
+                continue
+
+            connection_time = (flight.departure_time - current_time).total_seconds() / 3600
+            if connection_time < 1 or connection_time > 6:  # Connection time between 1 and 6 hours
+                continue
+
+            if flight.arrival_airport in visited:
+                continue
+
+            new_option = FlightOption()
+            new_option.flights = current_option.flights.copy()
+            new_option.add_flight(flight)
+
+            new_visited = visited.copy()
+            new_visited.add(flight.arrival_airport)
+
+            dfs(flight.arrival_airport, flight.arrival_time, new_option, new_visited)
+
+    initial_option = FlightOption()
+    dfs(departure_airport, departure_time, initial_option, {departure_airport})
+
+    return options
+
+def calculate_baseline_duration(options: List[FlightOption]) -> float:
+    direct_flights = [opt for opt in options if len(opt.flights) == 1]
+    if direct_flights:
+        return min(opt.total_duration for opt in direct_flights)
+   
+    min_connections = min(len(opt.flights) for opt in options)
+    least_connection_flights = [opt for opt in options if len(opt.flights) == min_connections]
+    return min(opt.total_duration for opt in least_connection_flights)
+
+def apply_discounts(options: List[FlightOption], baseline_duration: float):
+    durations = np.array([opt.total_duration for opt in options])
+    idle_times = np.array([opt.idle_time for opt in options])
+
+    duration_threshold = baseline_duration * 1.5
+    max_idle_time = np.max(idle_times)
+
+    for i, option in enumerate(options):
+        discount = 0
+        if durations[i] >= duration_threshold:
+            discount += option.total_price * 0.5
+            if idle_times[i] == max_idle_time and max_idle_time > 0:
+                discount += option.total_price * 0.2
+       
+        option.discount = min(discount, option.total_price * 0.7)  # Cap discount at 70%
+        option.total_price -= option.discount
+
+def calculate_ranking_scores(options: List[FlightOption], preferred_airline_code: str) -> List[FlightOption]:
+    prices = np.array([opt.total_price for opt in options])
+    durations = np.array([opt.total_duration for opt in options])
+    connections = np.array([len(opt.flights) - 1 for opt in options])
+    airline_matches = np.array([1 if any(flight.airline_code == preferred_airline_code for flight in opt.flights) else 0 for opt in options])
```

Comment:
This can be further improved. I'd love to see something like (the more you have traveled with this airline, the higher the score).

Feel free to implement that in future commit.

Reply:
This shouldn't be too hard. But might need some additional changes. When I add a flight to upcoming trip list, I can add the airline in an array and use it here to check if that's a frequently travelled option and assign higher score.